tools/interu/src/config/runner.rs

```diff
@@ -0,0 +1,174 @@
+
+        match parts[..] {
+            [distribution, version] => {
+                let distribution =
+                    Distribution::from_str(distribution).context(ParseDistributionSnafu)?;
+
+                Ok(PlatformPair {
+                    version: version.to_owned(),
+                    distribution,
+                })
+            }
+            _ => InvalidFormatSnafu.fail(),
+        }
+    }
+}
+
+impl Display for PlatformPair {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{distribution}-{version}",
+            distribution = self.distribution,
+            version = self.version,
+        )
+    }
+}
+
+#[derive(
+    Clone, Debug, Eq, Hash, PartialEq, Deserialize, Serialize, strum::Display, strum::EnumString,
+)]
+#[strum(serialize_all = "lowercase")]
+#[serde(rename_all = "lowercase")]
+pub enum Distribution {
+    Eks,
+    Gke,
+    Aks,
+    Kind,
+    K3s,
+    Rke2,
+}
+
+#[derive(
+    Clone, Debug, Eq, Hash, PartialEq, Deserialize, Serialize, strum::Display, strum::EnumString,
+)]
+#[strum(serialize_all = "kebab-case")]
+#[serde(rename_all = "kebab-case")]
+pub enum Architecture {
+    Amd64,
+    Arm64,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub enum Size {
+    Small,
+    Medium,
+    Large,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct NodeGroup {
+    name: String,
+
+    #[serde(rename = "arch")]
+    architecture: Architecture,
+
+    nodes: usize,
+    size: Size,
+    disk: usize,
+}
+
+#[derive(Debug, Snafu)]
+pub enum ConvertNodeGroupError {
+    #[snafu(display("unknown distribution {distribution}"))]
+    UnknownDistribution { distribution: Distribution },
+
+    #[snafu(display("unknown architecture {architecture}"))]
+    UnknownArchitecture { architecture: Architecture },
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct ReplicatedNodeGroup {
+    instance_type: String,
+    name: String,
+    nodes: usize,
+    disk: usize,
+}
+
+impl ReplicatedNodeGroup {
+    pub fn try_from(
+        node_group: NodeGroup,
+        instances: &Instances,
+        distribution: &Distribution,
+    ) -> Result<Self, ConvertNodeGroupError> {
+        let architectures =
+            instances
+                .get(distribution)
+                .with_context(|| UnknownDistributionSnafu {
+                    distribution: distribution.clone(),
+                })?;
+
+        let sizes = architectures
+            .get(&node_group.architecture)
+            .with_context(|| UnknownArchitectureSnafu {
+                architecture: node_group.architecture.clone(),
+            })?;
+
+        let instance_type = match node_group.size {
+            Size::Small => sizes.small.clone(),
+            Size::Medium => sizes.medium.clone(),
+            Size::Large => sizes.large.clone(),
```

Comment:
Just out of curiosity, do we need to clone here?
Can we consume `Instances` instead?

Reply:
I don't hink we can consume `Instances` here, because `ReplicatedNodeGroup::try_from` gets called repeately for each specified node group.

We can however change the `instance_type` field of `ReplicatedNodeGroup` from `String` to `&'a str` which allows us to capture the appropriate lifetime. In fact, every field can take references. Expect an upcoming commit.

Reply:
Done in c8089a0